cmdstanpy/stanfit.py

```diff
@@ -958,6 +965,61 @@ def draws_pd(
             return self._draws_pd
         return self._draws_pd[mask]
 
+    def draws_xr(
+        self, vars: List[str] = None, inc_warmup: bool = False
+    ) -> "xr.Dataset":
+        """
+        Returns the sampler draws as a xarray Dataset.
+        :param vars: optional list of variable names.
+        :param inc_warmup: When ``True`` and the warmup draws are present in
+            the output, i.e., the sampler was run with ``save_warmup=True``,
+            then the warmup draws are included.  Default value is ``False``.
+        """
+        if not XARRAY_INSTALLED:
+            raise RuntimeError(
+                "xarray is not installed, cannot produce draws array"
+            )
+        if inc_warmup and not self._save_warmup:
+            self._logger.warning(
+                "draws from warmup iterations not available,"
+                ' must run sampler with "save_warmup=True".'
+            )
+
+        if vars is None:
+            vars = self.stan_vars_dims.keys()
+
+        self._assemble_draws()
+
+        num_draws = self.num_draws_sampling
+        meta = self._metadata.cmdstan_config
+        attrs = {
+            "stan_version": f"{meta['stan_version_major']}."
+            f"{meta['stan_version_minor']}.{meta['stan_version_patch']}",
+            "model": meta["model"],
+            "num_unconstrained_params": self.num_unconstrained_params,
+            "num_draws_sampling": num_draws,
+        }
+        if inc_warmup and self._save_warmup:
+            num_draws += self.num_draws_warmup
+            attrs["num_draws_warmup"] = self.num_draws_warmup
+
+        data = {}
+        coordinates = {"chains": self.chain_ids, "draws": np.arange(num_draws)}
+        dims = ("draws", "chains")
+        for var in vars:
+            draw1 = 0
+            if not inc_warmup and self._save_warmup:
+                draw1 = self.num_draws_warmup
+            col_idxs = self._metadata.stan_vars_cols[var]
+            var_dims = dims + tuple(
+                f"{var}_dim_{i}" for i in range(len(self.stan_vars_dims[var]))
+            )
+            data[var] = (var_dims, np.squeeze(self._draws[draw1:, :, col_idxs]))
```

Comment:
What is this step doing? One needs to be really careful when using squeeze and usually, it doesn't work like it is supposed to.

Reply:
The first statement generates dimension names like "mu_dim_0", "mu_dim_1" if the stan variable has multiple dimensions. The squeeze exists because if there is only a single column index, the return of `self._draws[draw1:, :, col_idxs]` has shape (chains, dims, 1) but this is not desirable

Reply:
What happens when there is dimension size 1?

Reply:
You mean if either of chain or draws is size one? I suppose this could happen for chain. I will look into a refactor that avoids squeeze

Reply:
If there is a vector of size 1

Reply:
I think that would still show up as having a dimension in stan_var_dims, as opposed to the scalar values which have unit as their dimension, so it should be possible to handle the two separately. 

I think if forced to choose between all scalars having a phony dimension and length-1 vectors being reduced to scalars, I would pick the latter

Reply:
this is correct - we're parsing the column names, a scalar Stan variable name cannot contain '['.  if there is a Stan vector or array of length 1, it will have column name `foo[1]` and therefore will have a dimension.

Reply:
Ok, great.

So length 1 chain / draw cases needs to be resolved.

Reply:
Yes, I will refactor to accommodate